gwcs/wcs.py

```diff
@@ -285,40 +338,215 @@ def __call__(self, *args, **kwargs):
+               ``detect_divergence`` to `True` will incur about 5-10%
+               performance penalty with the larger penalty
+               corresponding to ``adaptive`` set to `True`.
+               Because the benefits of enabling this
+               feature outweigh the small performance penalty,
+               especially when ``adaptive`` = `False`, it is
+               recommended to set ``detect_divergence`` to `True`,
+               unless extensive testing of the distortion models for
+               images from specific instruments show a good stability
+               of the numerical method for a wide range of
+               coordinates (even outside the image itself).
+
+            .. note::
+               Indices of the diverging inverse solutions will be
+               reported in the ``divergent`` attribute of the
+               raised :py:class:`NoConvergence` exception object.
+
+        Returns
+        -------
+        result : tuple
+            Returns a tuple of values or arrays of values for each axis.
+
+        Raises
+        ------
+        NoConvergence
+            The iterative method did not converge to a
+            solution to the required accuracy within a specified
+            number of maximum iterations set by the ``maxiter``
+            parameter. To turn off this exception, set ``quiet`` to
+            `True`. Indices of the points for which the requested
+            accuracy was not achieved (if any) will be listed in the
+            ``slow_conv`` attribute of the
+            raised :py:class:`NoConvergence` exception object.
+
+            See :py:class:`NoConvergence` documentation for
+            more details.
+
+        NotImplementedError
+            Either iterative or analytic or both inverses have not been
+            implemented.
+
+        ValueError
+            Invalid argument values.
+
         """
         if not utils.isnumerical(args[0]):
             args = self.output_frame.coordinate_to_quantity(*args)
             if self.output_frame.naxes == 1:
                 args = [args]
-            if not self.backward_transform.uses_quantity:
-                args = utils.get_values(self.output_frame.unit, *args)
+            try:
+                if not self.backward_transform.uses_quantity:
+                    args = utils.get_values(self.output_frame.unit, *args)
+            except (NotImplementedError, KeyError):
+                pass
 
         with_units = kwargs.pop('with_units', False)
+        with_method = kwargs.pop('with_method', 'analytic')
+
         if 'with_bounding_box' not in kwargs:
             kwargs['with_bounding_box'] = True
+
         if 'fill_value' not in kwargs:
             kwargs['fill_value'] = np.nan
 
-        try:
-            result = self.backward_transform(*args, **kwargs)
-        except (NotImplementedError, KeyError):
-            result = self._invert(*args, **kwargs)
+        if with_method == 'analytic':
```

Comment:
Mihai and I talked for a bit about this issue. There are some longer term solutions we discussed at some length, but we decided that these shouldn't hold up this PR. So for now inverse should use the user inverse if it exists otherwise the iterative inverse if it doesn't. And yes, each of these should have a public interface for someone who specifically wants to use one or the other regardless. I'm a little fuzzy on what the public names should be. user_inverse & iter_inverse?

Reply:
How about `numerical_inverse`?